src/main/java/org/jboss/staxmapper/IntVersion.java

```diff
@@ -0,0 +1,172 @@
+     * @param segments the version segments
+     */
+    public IntVersion(int... values) {
+        // Defensive copy
+        this.segments = values.clone();
+    }
+
+    /**
+     * Returns the version segment at the specified index.
+     * @param index a version segment index
+     * @return the version segment at the specified index, or 0, if the index is out-of-bounds.
+     * @throws ArrayIndexOutOfBoundsException if index is negative
+     */
+    public int segment(int index) {
+        return (index < this.segments.length) ? this.segments[index] : 0;
+    }
+
+    /**
+     * Returns the number of segments with which this version was constructed.
+     * @return the number of segments with which this version was constructed.
+     */
+    public int segments() {
+        return this.segments.length;
+    }
+
+    /**
+     * Returns the 1st segment of this version.
+     * Equivalent to {@code this.segment(0)}.
+     * @return the 1st segment of this version.
+     */
+    public int major() {
+        return this.segment(0);
+    }
+
+    /**
+     * Returns the 2nd segment of this version.
+     * Equivalent to {@code this.segment(1)}.
+     * @return the 2nd segment of this version.
+     */
+    public int minor() {
+        return this.segment(1);
+    }
+
+    /**
+     * Returns the 3rd segment of this version.
+     * Equivalent to {@code this.segment(2)}.
+     * @return the 3rd segment of this version.
+     */
+    public int micro() {
+        return this.segment(2);
+    }
+
+    /**
+     * Returns a stream of the segments of this version.
+     * @return a stream of version segments
+     */
+    public IntStream stream() {
+        return IntStream.of(this.segments);
+    }
+
+    /**
+     * Prints this version using the specified segment delimiter, padded (or truncated) to the specified number of segments.
+     * @param a version segment delimiter
+     * @param the number of segments with which to pad this version's string representation
+     * @return a string representation of this version
+     */
+    public String toString(CharSequence delimiter, int segments) {
+        IntStream stream = (this.segments.length >= segments) ? IntStream.of(this.segments).limit(segments) : IntStream.range(0, segments).map(this::segment);
+        return stream.mapToObj(String::valueOf).collect(Collectors.joining(delimiter));
+    }
+
+    @Override
+    public int compareTo(IntVersion version) {
+        for (int i = 0; i < Math.max(this.segments.length, version.segments.length); ++i) {
+            int result = Integer.compare(this.segment(i), version.segment(i));
+            if (result != 0) return result;
+        }
+        return 0;
+    }
+
+    @Override
+    public int hashCode() {
+        // Ignore trailing zero segments
+        int length = this.segments.length;
+        while ((length > 0) && (this.segments[length -1] == 0)) {
+            length -= 1;
+        }
```

Comment:
It might be simpler to trim trailing zeros in the constructor. You could do this in the array constructor using `Arrays.copyOf(original, cnt)` to trim off as many zeros as you find.

Reply:
Indeed.  It also renders the fixed arg constructors redundant.
Fixed.